registry/remote/repository.go

```diff
@@ -944,37 +946,69 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
 }
 
 // generateDescriptor returns a descriptor generated from the response.
-func (s *manifestStore) generateDescriptor(resp *http.Response, ref registry.Reference) (ocispec.Descriptor, error) {
+func (s *manifestStore) generateDescriptor(resp *http.Response, ref registry.Reference, httpMethod string) (ocispec.Descriptor, error) {
+	// 1. Validate Content-Type
 	mediaType, _, err := mime.ParseMediaType(resp.Header.Get("Content-Type"))
 	if err != nil {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: invalid response Content-Type: %w", resp.Request.Method, resp.Request.URL, err)
+		return ocispec.Descriptor{}, fmt.Errorf(
+			"%s %q: invalid response `Content-Type` header; %w",
+			resp.Request.Method,
+			resp.Request.URL,
+			err,
+		)
 	}
 
+	// 2. Validate Size
 	size := resp.ContentLength
 	if size == -1 {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: unknown response Content-Length", resp.Request.Method, resp.Request.URL)
-	}
-
-	// validate digest if ref is a digest
-	if refDigest, err := ref.Digest(); err == nil {
-		if err = verifyContentDigest(resp, refDigest); err != nil {
-			return ocispec.Descriptor{}, err
+		return ocispec.Descriptor{}, fmt.Errorf(
+			"%s %q: unknown response Content-Length",
+			resp.Request.Method,
+			resp.Request.URL,
+		)
+	}
+
+	// 3. Validate Digest
+	var contentDigest, respDigest, refDigest, hdrDigest digest.Digest
+	var refDigestErr, hdrDigestErr error
+	if refDigest, err = ref.Digest(); err != nil {
+		refDigestErr = fmt.Errorf(
+			"%s %q: invalid reference digest: `%s`; %w",
+			resp.Request.Method,
+			resp.Request.URL,
+			respDigest,
+			err,
+		)
+	}
+	if hdrDigestStr := resp.Header.Get("Docker-Content-Digest"); hdrDigestStr != "" {
+		if hdrDigest, err = digest.Parse(hdrDigestStr); err != nil {
+			hdrDigestErr = fmt.Errorf(
+				"%s %q: invalid response header value `Docker-Content-Digest`: `%s`; %w",
+				resp.Request.Method,
+				resp.Request.URL,
+				hdrDigestStr,
+				err,
+			)
 		}
-		return ocispec.Descriptor{
-			MediaType: mediaType,
-			Digest:    refDigest,
-			Size:      size,
-		}, nil
 	}
 
-	digestStr := resp.Header.Get("Docker-Content-Digest")
-	if digestStr == "" {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: empty response Docker-Content-Digest", resp.Request.Method, resp.Request.URL)
-	}
+	// HTTP HEAD Requests are equipped with the `Docker-Content-Digest` header.
+	// HTTP GET Requests do not.
```

Comment:
HTTP GET Requests **may** not?
Some registries do return the `Docker-Content-Digest` header for Get requests.

Reply:
Also, can we add unit tests for the cases where the `Docker-Content-Digest` header is not returned?